### DISCUSSION TITLE
Select with Async data from api

### DISCUSSION THREAD

**Discussion by Nikhilchandra-Rathod** [mantinedev/mantine#4077] [Q&A]
Hello,

I want to make search functionality in select (It may have more than 100 results)
I tried it but I don't get event for onSelect option from select. (It call onSearchChange when type search term and also when we select option, so how we know its when select option but not by type)
and I want to immediately clear search select after select search term and put selected item in array.

Any one can help me in it.

here is the code for it
https://codesandbox.io/s/eager-drake-x78yzz?file=/src/App.tsx

I need it so please give me reply as soon as possible.

Thanks

**Comment by Nikhilchandra-Rathod** [comment]
here is updated codesandbox
https://codesandbox.io/s/aged-paper-3knp0m?file=/src/App.tsx
but I see here console.log printing multiple times, It should only once

**Comment by Nikhilchandra-Rathod** [comment]
here is updated codesendbox for more clarity
https://codesandbox.io/s/aged-paper-3knp0m?file=/src/App.tsx
Now, Items print in screen multiple times after select

@rtivital Any solution for it ? Please I need it very hardly so please reply me as soon as possible.

**Comment by teh23** [comment]
Why not MultiSelect? its good for this. Anyway, u can achieve this using onChange

**Comment by brycefranzen** [comment]
Main issues with this:
1. `onSearchChange` is triggered on input change AND when a value is set.
2. setting the initial form value doesn't guarantee that the select item exists in the data array

Here is how I resolved this on our project:
1. use a `ref` on the select input to watch for `keydown` which then triggers the update to our search params instead of relying on `onSearchChange`
2. added an additional `initialValueOptions` param which is filtered and added to the `data` param value
3. reset the search value when the clear button is pressed so all initial options are visible again

```tsx
type TAdditionalAsyncSelectProps = {
  /**
   * callback to set the search value used for async operation
   */
  setSearch: (newValue: string) => void;
  /**
   * Set to true of the async operation is loading
   */
  isLoading?: Boolean;
  /**
   * This is a hack to allow us to add additional options to the select.
   *
   * When you set the initial value of the select, you need to make sure that the
   * option is present in the data array. Without this, the select will be empty
   */
  initialValueOptions?: SelectItem[];
  /**
   * We never use `string` types for data, always the objects. Force this.
   */
  data: SelectItem[];
};


export const AsyncSelect = (props: Omit<SelectProps, 'data'> & TAdditionalAsyncSelectProps) => {
  const { setSearch, isLoading, data, initialValueOptions, ...restProps } = props;
  const selectInputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (selectInputRef.current) {
      const handleClientKeyDown = () => {
        setSearch(selectInputRef.current?.value ?? '');
      };
      selectInputRef.current.addEventListener('keydown', handleClientKeyDown);
    }
  }, [selectInputRef, setSearch]);

  const filteredInitialValues =
    initialValueOptions?.filter(
      (option) => data.findIndex((d) => d.value === option.value) === -1
    ) ?? [];

  return (
    <Select
      ref={selectInputRef}
      searchable
      data={[...data, ...filteredInitialValues]}
      clearButtonProps={{
        onClickCapture: () => {
          setSearch('');
        },
      }}
      clearable
      rightSection={isLoading ? <Loader size="xs" /> : undefined}
      {...restProps}
    />
  );
};

/** USAGE */
const [agenciesSearch, setAgenciesSearch] = useDebouncedState('', 300);

<AsyncSelect
  label="Agency"
  setSearch={setAgenciesSearch}
  nothingFound="No agencies found"
  data={
    agencySelect?.agencies.items?.map((agency) => ({
      label: agency.name,
      value: agency.id,
    })) ?? []
  }
  initialValueOptions={
    initialValues
      ? [
          {
            label: initialValues.agency.name,
            value: initialValues.agency.id,
          },
        ]
      : []
  }
  required
  isLoading={agencySelectLoading}
  {...getInputProps('agencyId')}
/>
```
@Nikhilchandra-Rathod @teh23 Hope this helps!
